src/libraries/System.Linq/src/System/Linq/ToCollection.cs

```diff
@@ -25,22 +20,19 @@ public static TSource[] ToArray<TSource>(this IEnumerable<TSource> source)
 
             if (source is ICollection<TSource> collection)
             {
-                int count = collection.Count;
-                if (count != 0)
-                {
-                    var result = new TSource[count];
-                    collection.CopyTo(result, 0);
-                    return result;
-                }
-
-                return [];
+                return ICollectionToArray(collection);
             }
 
             return EnumerableToArray(source);
 
             [MethodImpl(MethodImplOptions.NoInlining)] // avoid large stack allocation impacting other paths
             static TSource[] EnumerableToArray(IEnumerable<TSource> source)
             {
+                if (source is null)
```

Comment:
Nit: nesting validation inside the local method will result in its compiler-generated name to show up in the stack trace. I would prefer it if validation was done separately on each public method even if that results in duplication.